substrate/utils/frame/benchmarking-cli/src/overhead/cmd.rs

```diff
@@ -92,18 +183,333 @@ pub(crate) enum BenchmarkType {
+/// inherents are required for every chain. The runtime will pick the ones
+/// it requires based on their identifier.
+fn create_inherent_data<Client: UsageProvider<Block> + HeaderBackend<Block>, Block: BlockT>(
+	client: &Arc<Client>,
+	chain_type: &ChainType,
+) -> InherentData {
+	let genesis = client.usage_info().chain.best_hash;
+	let header = client.header(genesis).unwrap().unwrap();
+
+	let mut inherent_data = sp_inherents::InherentData::new();
+
+	// Para inherent can only makes sense when we are handling a parachain.
+	if let Parachain(para_id) = chain_type {
+		let parachain_validation_data_provider = MockValidationDataInherentDataProvider::<()> {
+			para_id: ParaId::from(*para_id),
+			current_para_block_head: Some(header.encode().into()),
+			relay_offset: 1,
+			..Default::default()
+		};
+		let _ = futures::executor::block_on(
+			parachain_validation_data_provider.provide_inherent_data(&mut inherent_data),
+		);
+	}
+
+	// Parachain inherent that is used on relay chains to perform parachain validation.
+	let para_inherent = polkadot_primitives::InherentData {
+		bitfields: Vec::new(),
+		backed_candidates: Vec::new(),
+		disputes: Vec::new(),
+		parent_header: header,
+	};
+
+	// Timestamp inherent that is very common in substrate chains.
+	let timestamp = sp_timestamp::InherentDataProvider::new(std::time::Duration::default().into());
+
+	let _ = futures::executor::block_on(timestamp.provide_inherent_data(&mut inherent_data));
+	let _ =
+		inherent_data.put_data(polkadot_primitives::PARACHAINS_INHERENT_IDENTIFIER, &para_inherent);
+
+	inherent_data
+}
+
+/// Identifies what kind of chain we are dealing with.
+///
+/// Chains containing the `ParachainSystem` and `ParachainInfo` pallet are considered parachains.
+/// Chains containing the `ParaInherent` pallet are considered relay chains.
+fn identify_chain(metadata: &Metadata, para_id: Option<u32>) -> ChainType {
+	let parachain_info_exists = metadata.pallet_by_name("ParachainInfo").is_some();
+	let parachain_system_exists = metadata.pallet_by_name("ParachainSystem").is_some();
+	let para_inherent_exists = metadata.pallet_by_name("ParaInherent").is_some();
+
+	log::debug!("{} ParachainSystem", if parachain_system_exists { "✅" } else { "❌" });
+	log::debug!("{} ParachainInfo", if parachain_info_exists { "✅" } else { "❌" });
+	log::debug!("{} ParaInherent", if para_inherent_exists { "✅" } else { "❌" });
+
+	let chain_type = if parachain_system_exists && parachain_info_exists {
+		Parachain(para_id.unwrap_or(DEFAULT_PARA_ID))
+	} else if para_inherent_exists {
+		Relaychain
+	} else {
+		Unknown
+	};
+
+	log::info!(target: LOG_TARGET, "Identified Chain type from metadata: {}", chain_type);
+
+	chain_type
+}
+
+#[derive(Deserialize, Serialize, Clone, ChainSpecExtension)]
+pub struct ParachainExtension {
+	/// The id of the Parachain.
+	pub para_id: Option<u32>,
+}
+
 impl OverheadCmd {
+	fn state_handler_from_cli<HF: HostFunctions>(
```

Comment:
Maybe this logic could be moved to some shared function and re-used in pallet cmd?

Reply:
I thought about that and experimented a bit, but decided against that.
Pallet and overhead handle things just slightly different, pallet does not care about `para_id` at all, overhead does not accept `GenesisBuilder::None`. These subtle differences are a bit annoying and would create additional control flow. So I went with this slight duplication.